File: minion/app/src/main/java/snakesoft/minion/CalendarItemViewActivity.java
package snakesoft.minion;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import calendar.Calendar;
import common.Common;

public class CalendarItemViewActivity extends AppCompatActivity implements View.OnClickListener {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_calendar_item_view);

        TextView txtView = (TextView) findViewById(R.id.cal_txtText);
        txtView.setText(getIntent().getStringExtra("Text"));

        TextView txtViewTime = (TextView) findViewById(R.id.cal_txtTextTime);
        txtViewTime.setText(getTimeText(getIntent().getLongExtra("Timestamp", 0)));

        TextView txtViewDate = (TextView) findViewById(R.id.cal_txtTextDate);
        txtViewDate.setText(getDateText(getIntent().getLongExtra("Timestamp", 0)));

        mAcceptBtn = (Button) findViewById(R.id.cal_btnAccept);
        mAcceptBtn.setOnClickListener(CalendarItemViewActivity.this);

        mBtnDelete = (Button) findViewById(R.id.cal_btnDelete);
        mBtnDelete.setOnClickListener(CalendarItemViewActivity.this);

        if(getIntent().getBooleanExtra("IsNewItem", false))
        {
            mBtnDelete.setVisibility(View.INVISIBLE);
        }
    }

    private String getDateText(long timestamp)
    {
        java.util.Calendar cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"));
        cal.setTimeInMillis(timestamp * 1000);

        String days = Integer.toString(cal.get(java.util.Calendar.DAY_OF_MONTH));
        if(days.length() == 1)
        {
            days = "0" + days;
        }

        String month = Integer.toString(cal.get(java.util.Calendar.MONTH) + 1);
        if(month.length() == 1)
        {
            month = "0" + month;
        }

        return days + "/" + month + "/" + Integer.toString(cal.get(java.util.Calendar.YEAR));
    }

    private String getTimeText(long timestamp)
    {
        java.util.Calendar cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"));
        cal.setTimeInMillis(timestamp * 1000);

        String minutes = Integer.toString(cal.get(java.util.Calendar.MINUTE));
        if(minutes.length() == 1)
        {
            minutes = "0" + minutes;
        }

        return Integer.toString(cal.get(java.util.Calendar.HOUR_OF_DAY)) + ":" +
                minutes;
    }

    @Override
    public void onClick(View v)
    {
        final String guid = getIntent().getStringExtra("Id");
        if(v == mAcceptBtn)
        {
            TextView txtView = (TextView) findViewById(R.id.cal_txtText);
            TextView txtViewTime = (TextView) findViewById(R.id.cal_txtTextTime);
            TextView txtViewDate = (TextView) findViewById(R.id.cal_txtTextDate);

            long timestamp = parseTimeAndDateText(
                    txtViewTime.getText().toString(),
                    txtViewDate.getText().toString());

            if(timestamp == 0)
            {
                Context context = getApplicationContext();
                CharSequence text = "Error in parsing text/date fields";
                int duration = Toast.LENGTH_SHORT;

                Toast toast = Toast.makeText(context, text, duration);
                toast.show();
                return;
            }

            Calendar.CalendarItem result = Calendar.CalendarItem.newBuilder()
                    .setText(txtView.getText().toString())
                    .setTimestamp(timestamp)
                    .setId(Common.UniqueId.newBuilder().setGuid(guid).build())
                    .build();

            if(getIntent().getBooleanExtra("IsNewItem", false))
            {
                GlobalModel.getCalendarModel().addItem(result);
            }
            else
            {
                GlobalModel.getCalendarModel().modifyItem(result);
            }
            finish();
        }
        if(v == mBtnDelete)
        {
            AlertDialog.Builder dlgAlert  = new AlertDialog.Builder(this);
            dlgAlert.setMessage("Are you sure?");
            dlgAlert.setTitle("Caution");
            dlgAlert.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int which) {
                    GlobalModel.getCalendarModel().deleteItem(guid);
                    finish();
                }});

            dlgAlert.setCancelable(true);
            dlgAlert.create().show();
        }
    }

    private long parseTimeAndDateText(String timeS, String dateS)
    {
        try
        {
            java.util.Calendar cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"));
            String[] values = timeS.split(":");

            int hours = Integer.valueOf(values[0]);
            int minutes = Integer.valueOf(values[1]);

            values = dateS.split("/");

            int day = Integer.valueOf(values[0]);
            int month = Integer.valueOf(values[1]) - 1;
            int year = Integer.valueOf(values[2]);

            cal.set(year, month, day, hours, minutes);

            return cal.getTimeInMillis() / 1000;
        }
        catch (Exception ex)
        {
            return 0;
        }
    }

    private Button mBtnDelete;
    private Button mAcceptBtn;
}

File: minion/app/src/main/java/snakesoft/minion/Activities/CalendarItemViewActivity.kt
package snakesoft.minion.Activities

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast

import calendar.Calendar
import common.Common
import snakesoft.minion.Models.CalendarItem
import snakesoft.minion.Models.GlobalModel
import snakesoft.minion.R

class CalendarItemViewActivity : AppCompatActivity(), View.OnClickListener {

    private var mBtnDelete: Button? = null
    private var mAcceptBtn: Button? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_calendar_item_view)

        val txtView = findViewById(R.id.cal_txtText) as TextView
        txtView.text = intent.getStringExtra("Text")

        val txtViewTime = findViewById(R.id.cal_txtTextTime) as TextView
        txtViewTime.text = getTimeText(intent.getLongExtra("Timestamp", 0))

        val txtViewDate = findViewById(R.id.cal_txtTextDate) as TextView
        txtViewDate.text = getDateText(intent.getLongExtra("Timestamp", 0))

        mAcceptBtn = findViewById(R.id.cal_btnAccept) as Button
        mAcceptBtn!!.setOnClickListener(this@CalendarItemViewActivity)

        mBtnDelete = findViewById(R.id.cal_btnDelete) as Button
        mBtnDelete!!.setOnClickListener(this@CalendarItemViewActivity)

        if (intent.getBooleanExtra("IsNewItem", false)) {
            mBtnDelete!!.visibility = View.INVISIBLE
        }
    }

    private fun getDateText(timestamp: Long): String {
        val cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"))
        cal.timeInMillis = timestamp * 1000

        var days = Integer.toString(cal.get(java.util.Calendar.DAY_OF_MONTH))
        if (days.length == 1) {
            days = "0$days"
        }

        var month = Integer.toString(cal.get(java.util.Calendar.MONTH) + 1)
        if (month.length == 1) {
            month = "0$month"
        }

        return days + "/" + month + "/" + Integer.toString(cal.get(java.util.Calendar.YEAR))
    }

    private fun getTimeText(timestamp: Long): String {
        val cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"))
        cal.timeInMillis = timestamp * 1000

        var minutes = Integer.toString(cal.get(java.util.Calendar.MINUTE))
        if (minutes.length == 1) {
            minutes = "0$minutes"
        }

        return Integer.toString(cal.get(java.util.Calendar.HOUR_OF_DAY)) + ":" +
                minutes
    }

    override fun onClick(v: View) {
        val guid = java.util.UUID.fromString(intent.getStringExtra("Id"))
        if (v === mAcceptBtn) {
            val txtView = findViewById(R.id.cal_txtText) as TextView
            val txtViewTime = findViewById(R.id.cal_txtTextTime) as TextView
            val txtViewDate = findViewById(R.id.cal_txtTextDate) as TextView

            val timestamp = parseTimeAndDateText(
                    txtViewTime.text.toString(),
                    txtViewDate.text.toString())

            if (timestamp == 0L) {
                val context = applicationContext
                val text = "Error in parsing text/date fields"
                val duration = Toast.LENGTH_SHORT

                val toast = Toast.makeText(context, text, duration)
                toast.show()
                return
            }

            val result = CalendarItem(guid, txtView.text.toString(), timestamp);

            if (intent.getBooleanExtra("IsNewItem", false)) {
                GlobalModel.calendarModel!!.addItem(result)
            } else {
                GlobalModel.calendarModel!!.replaceItem(result)
            }
            finish()
        }
        if (v === mBtnDelete) {
            val dlgAlert = AlertDialog.Builder(this)
            dlgAlert.setMessage("Are you sure?")
            dlgAlert.setTitle("Caution")
            dlgAlert.setPositiveButton("OK") { dialog, which ->
                GlobalModel.calendarModel!!.deleteItem(guid)
                finish()
            }

            dlgAlert.setCancelable(true)
            dlgAlert.create().show()
        }
    }

    private fun parseTimeAndDateText(timeS: String, dateS: String): Long {
        try {
            val cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"))
            var values = timeS.split(":".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

            val hours = Integer.valueOf(values[0])
            val minutes = Integer.valueOf(values[1])

            values = dateS.split("/".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()

            val day = Integer.valueOf(values[0])
            val month = Integer.valueOf(values[1]) - 1
            val year = Integer.valueOf(values[2])

            cal.set(year, month, day, hours, minutes)

            return cal.timeInMillis / 1000
        } catch (ex: Exception) {
            return 0
        }

    }
}
